### ISSUE TITLE
Entire CJK stack upside down

### ISSUE THREAD

**Issue by alerque** [sile-typesetter/sile#1245]
I'm monkey-patching some stuff through to get the manual to build for v0.12.0, but the entire Japanese language support and Tate / Ruby package scene is chaos. I'm adding fallback fonts early everywhere so the units and things work if called from anything other than an already working environment, but the order is all wrong. You can't measure glyphs before you load fonts, which means you can't measure glyphs in class init.

**Comment by alerque** [comment]
Basically see everything in #1215 except the first commit, particularly d76fd7b..ceba809.

**Comment by ctrlcctrlv** [comment]
I totally agree with you. I have a better version of the Ruby package which I will be PR'ing _soon_…it's just a matter of finding a charger, fixing an old laptop etc. Maybe I'll get to it now. But yes, really `tate.lua`/`ruby.lua` is not so good but I find myself quite outnumbered in terms of caring about doing anything as I seem to be the only user. Furthermore, I only am interested in Japanese typesetting (so the J in CJK) making me an unsuitable general expert.

**Comment by Omikhleia** [comment]
I don't know if it fits here in this issue, but there seem to be some code smell in the `tate`/`break-firstfit`/`ja` logic.

- `languages/ja.lua` [marks some nodes](https://github.com/sile-typesetter/sile/blob/2888d6c6cc931dae31b72cd395d8cfef898a1a8f/languages/ja.lua#L163) as "hangable", it seems to be the only occurrence in the sources.
- `packages/break-firstfit.lua` [has a test](https://github.com/sile-typesetter/sile/blob/2888d6c6cc931dae31b72cd395d8cfef898a1a8f/packages/break-firstfit/init.lua#L14) on `isHangable`, also a unique occurrence in the sources
- `packages/tate` [loads and uses the firstfit method](https://github.com/sile-typesetter/sile/blob/2888d6c6cc931dae31b72cd395d8cfef898a1a8f/packages/tate/init.lua#L15)

I am illiterate as far as Japanese is concerned, and have know idea what a "hangable" thing here in this language, but anyhow I can't see how the `isHangable` would ever be valued. It doesn't seem we have some logic for it (as opposed, say, to the `is_box` etc. constructs, which are valued via the node's type at construction).

My apologies if the above assumption is wrong -- It's just something I saw by accident and which left me perplexed.

**Comment by ctrlcctrlv** [comment]
@Omikhleia It should _not_ be removed. The "hangable" there refers to characters that may be in the ぶら下げ組 (_burasage-gumi_; lit. _(maybe) overhanging class (of characters)_) of jlreq §2.5.1.c, which states:

> [**ぶら下げ組**](https://www.w3.org/TR/jlreq/#term.line-adjustment-by-hanging-punctuation)とよばれる処理をした場合は，[行頭禁則](https://www.w3.org/TR/jlreq/#term.line-start-prohibition-rule)の処理を必要とする[句点類（cl-06）](https://www.w3.org/TR/jlreq/#cl-06)及び[読点類（cl-07）](https://www.w3.org/TR/jlreq/#cl-07)に限り，[行末](https://www.w3.org/TR/jlreq/#term.line-end)の版面の領域の外側に接して配置する（[Figure 38](https://www.w3.org/TR/jlreq/#fig_ad1_7)）．なお，ぶら下げ組についてはJIS X 4051に規定されていないが，同規格の解説8.1）c）に説明がある．

> [Line adjustment by **hanging punctuation**](https://www.w3.org/TR/jlreq/#term.line-adjustment-by-hanging-punctuation) is only necessary for [full stops (cl-06)](https://www.w3.org/TR/jlreq/#cl-06) and [commas (cl-07)](https://www.w3.org/TR/jlreq/#cl-07) when they would otherwise need to [be wrapped to the line head](https://www.w3.org/TR/jlreq/#term.line-start-prohibition-rule). The character is placed so that it touches the hanmen at the [line end](https://www.w3.org/TR/jlreq/#term.line-end) (see [Figure 38](https://www.w3.org/TR/jlreq/#fig_ad1_7)). (Hanging punctuation is not defined in JIS X 4051, but there is an explanation in sec. 8.1, c.)

_Cf._ <https://www.w3.org/TR/jlreq/#id83>

![image](https://user-images.githubusercontent.com/838783/175439576-64ed1482-9dce-434b-b552-67c8ce0681da.png)

It should be fixed. `isHangable` and `hangable` are identical entities. This being broken is highly unsurprising given the sad state of `tate` (although it's better than nothing which is what most software that attempts to compete with LaTeX provides).

**Comment by ctrlcctrlv** [comment]
Btw I suggest moving a lot of `languages/ja.lua` into a new file to be called `packages/jlreq.lua`. There are reasons you'd want some jlreq algorithms for non-Japanese. @alerque ?

**Comment by alerque** [comment]
Yes, any tooling that is generic for a family of genre of languages could be abstracted to module files that can be `require()`ed into whatever specific languages they are needed. Eventually I'll probably make the language support modules actually classes so they have an official inheritance model to follow so we can deal with region codes etc., but that isn't off the ground yet. For now my only comment on the proliferation of Lua files is that *all* files to be loaded with `require()` should return modules or functions rather than doing anything directly. They can have stuff in their local scope if desired but loading them should not touch anything in the global scope until explicitly called. We have a lot of code that wasn't written that way and it's costing me a lot of time to fix along the way.

**Comment by ctrlcctrlv** [comment]
@simoncozens seems to be on leave from SILE but I would really like his comment on the patch I put on the other issue (https://github.com/sile-typesetter/sile/issues/501#issuecomment-1165524715) against ctrlcctrlv/sile@14b7547cb25b4b1dc337cdcd4e8b291582eaf556.

**Comment by alerque** [comment]
So, um, GitHub is *very* bad at surfacing notifications for some things. @ctrlcctrlv I was just mucking around looking for how to set an away message since I'm going to be mostly off the computer for a few days, and I spotted something on my profile I missed for more than a month. More on that later when I'm back I suppose, I just wanted to at least start with a duly deserved “dankon”!
